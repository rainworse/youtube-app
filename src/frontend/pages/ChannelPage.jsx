import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import APIFetcher from '../../backend/APIFetcher';
import '../styles/ChannelPage.css';
import ConfigHelper from '../../backend/ConfigHelper';
import VideoCard from '../components/VideoCard';

const ChannelPage = () => {
  const params = useParams();
  const channelVideoCount = ConfigHelper.getValue('channel-page-video-count');
  const [channelData, setChannelData] = useState(null);
  const [channelVideoData, setChannelVideoData] = useState(null);

  useEffect(() => {
    const fetchChannelData = async () => {
      const channelData = await APIFetcher.fetchChannelData(params.id);
      setChannelData(channelData);
    };
    const fetchChannelVideoData = async () => {
      const videoData = await APIFetcher.fetchChannelVideos(
        params.id,
        channelVideoCount
      );
      setChannelVideoData(videoData);
    };

    fetchChannelData();
    fetchChannelVideoData();
  }, []);

  return (
    <div className="channel-page">
      {channelData === null ? null : (
        <div className="channel-info">
          <img
            className="channel-thumbnail"
            src={channelData.thumbnail}
            alt={channelData.title}
          />
          <div className="channel-text-info">
            <div className="channel-subscriber-count">
              Subscribers: {channelData.subscriberCount}
            </div>
            <div className="channel-title">{channelData.title}</div>
          </div>
        </div>
      )}
      <div className="channel-videos">
        {channelVideoData === null
          ? null
          : channelVideoData.map((v) => {
              return <VideoCard data={v} key={v.id} />;
            })}
      </div>
    </div>
  );
};

export default ChannelPage;
