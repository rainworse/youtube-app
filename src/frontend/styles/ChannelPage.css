.channel-page {
  width: 100%;
}

.channel-info {
  margin-top: 50px;
  margin-left: 50px;
  display: flex;
  align-items: end;
}

.channel-thumbnail {
  width: 150px;
  border-radius: 15px;
}

.channel-text-info {
  margin-left: 30px;
  margin-right: 30px;
}

.channel-title {
  font-size: 30px;
  font-weight: bold;
}

.channel-videos {
  margin-top: 50px;
  margin-left: 50px;
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
}
