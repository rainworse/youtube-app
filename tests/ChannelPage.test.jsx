import { render, waitFor } from '@testing-library/react';
import { expect, vi } from 'vitest';
import ChannelPage from '../src/frontend/pages/ChannelPage';
import UserContext from '../src/frontend/UserContext';
import { MemoryRouter } from 'react-router-dom';

vi.mock('../src/backend/APIFetcher');

describe('ChannelPage', () => {
  it('renders channel page', async () => {
    const channelPage = render(
      <MemoryRouter>
        <ChannelPage />
      </MemoryRouter>
    );

    await waitFor(() => channelPage.getByText('test channel'));
    await waitFor(() => channelPage.getByText('test video3'));

    expect(channelPage.container).toMatchSnapshot();
  });
});
