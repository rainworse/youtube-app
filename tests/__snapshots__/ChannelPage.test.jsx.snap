// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChannelPage > renders channel page 1`] = `
<div>
  <div
    class="channel-page"
  >
    <div
      class="channel-info"
    >
      <img
        alt="test channel"
        class="channel-thumbnail"
        src="fake/channel/thumbnail"
      />
      <div
        class="channel-text-info"
      >
        <div
          class="channel-subscriber-count"
        >
          Subscribers: 
          666
        </div>
        <div
          class="channel-title"
        >
          test channel
        </div>
      </div>
    </div>
    <div
      class="channel-videos"
    >
      <div
        class="video-card"
      >
        <a
          class="video-card-thumbnail-anchor"
          href="/video/001"
        >
          <img
            alt="test video1"
            class="video-card-thumbnail"
            src="non/existent/location1"
          />
          <div
            class="video-card-text"
          >
            <div
              class="video-card-title"
            >
              test video1
            </div>
            <a
              class="video-card-channel-anchor"
              href="/channel/channel001"
            >
              <div
                class="video-card-channel"
              >
                by 
                test channel1
              </div>
            </a>
          </div>
        </a>
      </div>
      <div
        class="video-card"
      >
        <a
          class="video-card-thumbnail-anchor"
          href="/video/002"
        >
          <img
            alt="test video2"
            class="video-card-thumbnail"
            src="non/existent/location2"
          />
          <div
            class="video-card-text"
          >
            <div
              class="video-card-title"
            >
              test video2
            </div>
            <a
              class="video-card-channel-anchor"
              href="/channel/channel002"
            >
              <div
                class="video-card-channel"
              >
                by 
                test channel2
              </div>
            </a>
          </div>
        </a>
      </div>
      <div
        class="video-card"
      >
        <a
          class="video-card-thumbnail-anchor"
          href="/video/003"
        >
          <img
            alt="test video3"
            class="video-card-thumbnail"
            src="non/existent/location3"
          />
          <div
            class="video-card-text"
          >
            <div
              class="video-card-title"
            >
              test video3
            </div>
            <a
              class="video-card-channel-anchor"
              href="/channel/channel003"
            >
              <div
                class="video-card-channel"
              >
                by 
                test channel3
              </div>
            </a>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
